src/TestExplorer/TestDiscovery.ts

```diff
@@ -148,11 +165,29 @@ function upsertTestItem(
         newItem = existingItem;
     }
 
+    // At this point all the test items that should have been deleted are out of the tree.
+    // Its possible we're dropping a whole branch of test items on top of an existing one,
+    // and we want to merge these branches instead of the new one replacing the existing one.
+    if (existingItem) {
+        deepMergeTestItemChildren(existingItem, newItem);
+    }
+
     // Manually add the test style as a tag so we can filter by test type.
     newItem.tags = [{ id: testItem.style }, ...testItem.tags];
     newItem.label = testItem.label;
     newItem.range = testItem.location?.range;
 
+    // Sort test items by path:line_number to provide a stable sort order.
+    // This prevents test items from changing their order as items are removed and readded.
+    if (testItem.location) {
```

Comment:
Is this necessary? In the UI it sorts them at least by line number once you set the location.